implants/lib/eldritch/src/pivot/arp_scan_impl.rs

```diff
@@ -1,5 +1,199 @@
+            return;
+        }
+        let now = SystemTime::now();
+        loop {
+            if now.elapsed().unwrap() > Duration::from_secs(5) {
+                break;
+            }
+            match rx.next() {
+                Ok(packet) => {
+                    let eth = match EthernetPacket::new(packet) {
+                        Some(eth) => eth,
+                        None => continue,
+                    };
+                    let arp = match ArpPacket::new(eth.payload()) {
+                        Some(arp) => arp,
+                        None => continue,
+                    };
+                    if arp.get_operation() != ArpOperations::Reply {
+                        continue;
+                    }
+                    let source_ip = arp.get_sender_proto_addr();
+                    let source_mac = arp.get_sender_hw_addr();
+                    if source_ip == ip {
+                        *data.lock().unwrap().get_mut(&ip).unwrap() = Some(ArpResponse {
+                            _source_ip: source_ip,
+                            source_mac,
+                            interface: interface.name.clone(),
+                        });
+                        break;
+                    }
+                }
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::TimedOut {
+                        continue;
+                    }
+                    println!("Error receiving packet: {}", e);
+                    return;
+                }
+            }
+        }
+    }
+}
+
+pub fn arp_scan(starlark_heap: &Heap, target_cidrs: Vec<String>) -> Result<Vec<Dict>> {
+    let target_cidrs = target_cidrs
+        .iter()
+        .map(|cidr| {
+            let (addr, prefix) = cidr.split_at(cidr.find('/').unwrap());
+            let addr = match Ipv4Addr::from_str(addr) {
+                Ok(addr) => addr,
+                Err(_) => {
+                    return Err(anyhow::anyhow!("Invalid IPv4 address: {}", addr));
+                }
+            };
+            let prefix: u8 = match prefix[1..].parse() {
+                Ok(prefix) => prefix,
+                Err(_) => {
+                    return Err(anyhow::anyhow!("Invalid CIDR prefix: {}", prefix));
+                }
+            };
+            let network = match Ipv4Network::new(addr, prefix) {
+                Ok(network) => network,
+                Err(_) => {
+                    return Err(anyhow::anyhow!("Invalid CIDR: {}", cidr));
+                }
+            };
+            Ok(network)
+        })
+        .collect::<Result<Vec<Ipv4Network>>>();
+    match target_cidrs {
+        Ok(target_cidrs) => {
+            let listener_out: Arc<Mutex<HashMap<Ipv4Addr, Option<ArpResponse>>>> =
+                Arc::new(Mutex::new(HashMap::new()));
+            for target_cidr in target_cidrs {
+                for ip in target_cidr.iter() {
+                    listener_out.lock().unwrap().insert(ip, None);
```

Comment:
Please avoid using unwrap in production code.
It’s better to either bubble up with a question mark or in this case since we’re in the root function handle the error returning a verbose anyhow error like “apr_scan: failed to”